firedrake/assemble.py

```diff
@@ -169,239 +160,240 @@ def thunk():
-    representing the result. This will be a :class:`float` for 0-forms/rank-0
-    Slate tensors, a :class:`.Function` for 1-forms/rank-1 Slate tensors and
-    a :class:`.Matrix` for 2-forms/rank-2 Slate tensors.
+def get_matrix(form, mat_type, sub_mat_type, *, bcs=None,
+               options_prefix=None, tensor=None, appctx=None,
+               form_compiler_parameters=None):
+    matfree = mat_type == "matfree"
+    arguments = form.arguments()
+    if tensor is not None and tensor.a.arguments() != arguments:
+        raise ValueError("Form's arguments do not match provided result tensor")
+    if matfree:
+        if tensor is None:
+            tensor = matrix.ImplicitMatrix(form, bcs,
+                                           fc_params=form_compiler_parameters,
+                                           appctx=appctx,
+                                           options_prefix=options_prefix)
+        elif not isinstance(tensor, matrix.ImplicitMatrix):
+            raise ValueError("Expecting implicit matrix with matfree")
+        else:
+            pass
+        return tensor, (), lambda: tensor
 
-    :arg bcs: A tuple of :class`.DirichletBC`\s and/or :class`.EquationBCSplit`\s to be applied.
-    :arg tensor: An existing tensor object into which the form should be
-        assembled. If this is not supplied, a new tensor will be created for
-        the purpose.
-    :arg form_compiler_parameters: (optional) dict of parameters to pass to
-        the form compiler.
-    :arg inverse: (optional) if f is a 2-form, then assemble the inverse
-         of the local matrices.
-    :arg mat_type: (optional) type for assembled matrices, one of
-        "nest", "aij", "baij", or "matfree".
-    :arg sub_mat_type: (optional) type for assembled sub matrices
-        inside a "nest" matrix.  One of "aij" or "baij".
-    :arg appctx: Additional information to hang on the assembled
-         matrix if an implicit matrix is requested (mat_type "matfree").
-    :arg options_prefix: An options prefix for the PETSc matrix
-        (ignored if not assembling a bilinear form).
-    """
-    if mat_type is None:
-        mat_type = parameters.parameters["default_matrix_type"]
-    if mat_type not in ["matfree", "aij", "baij", "nest", "dense"]:
-        raise ValueError("Unrecognised matrix type, '%s'" % mat_type)
-    if sub_mat_type is None:
-        sub_mat_type = parameters.parameters["default_sub_matrix_type"]
-    if sub_mat_type not in ["aij", "baij"]:
-        raise ValueError("Invalid submatrix type, '%s' (not 'aij' or 'baij')", sub_mat_type)
+    if isinstance(form, slate.TensorBase):
+        # FIXME: inherit from slate form somehow.
+        integral_types = {"cell"}
+    else:
+        integral_types = set(i.integral_type() for i in form.integrals())
+        for bc in bcs:
+            integral_types.update(integral.integral_type()
+                                  for integral in bc.integrals())
```

Comment:
@ksagiyam I migrated this from the previous assembly code. Why can we not apply equation bcs to slate tensors? Moreover, if this is a technical limitation, I suppose that one should raise a `NotImplementedError` in the slate case.